### PR TITLE
add UTC full support

### PR DESCRIPTION
**Pull request by saintplay**
Hello,
I just noticed that this component always handle dates in localzone time, which is super inconvenient for absolute timestamps (UTC).

**Related issues**: #23, #200, #313, #158, #118, etc

So I added a **non-breaking change** to use **UTC** for time calculations.
This option will be available via the prop **use-utc**

Looking for your thoughts on this,
Regards

### CODE REVIEW

**Comment by saintplay** [comment]
@charliekassel I found **important** issues in the test suite

**Comment by saintplay** [comment]
For anyone interested in: there is a **npm package published**

```bash
npm install --save @4keys/vuejs-datepicker
```

and

```js
import DatePicker from '@4keys/vuejs-datepicker';
```

**Comment by charliekassel** [comment]
Hey @saintplay - your contribution here is appreciated and I thank you for it.
I'm just a little hesitant to merge it without spending some time looking through it. It has not been forgotten though so please bear with me.

**Comment by saintplay** [comment]
Yeah I know this represent a noticeable refactoring.

**Under the hood I did the following:**

<br/>

- Added a `use-utc` prop

<br/>

- Use `getUTCFullYear` instead of `getFullYear` if `use-utc` prop is true
- Use `getUTCMonth` instead of `getMonth` if `use-utc` prop is true
- Use `getUTCDate` instead of `getDate` if `use-utc` prop is true
- Use `getUTCDay` instead of `getDay` if `use-utc` prop is true
- Use `getUTCHours` instead of `getHours` if `use-utc` prop is true
- Use `getUTCMinutes` instead of `getMinutes` if `use-utc` prop is true

<br/>

- Use `setUTCFullYear` instead of `setFullYear` if `use-utc` prop is true
- Use `setUTCMonth` instead of `setMonth` if `use-utc` prop is true
- Use `setUTCDate` instead of `setDate` if `use-utc` prop is true

<br/>

- Add method to compare Dates(without time) effectively: `this.compareDates(date1, date2)`

**Comment by saintplay** [comment]
**The following PR would (likely) break this feature**

- [ ] #390
- [ ] #430 

So we have two options

- **Wait for those PR to be resolved, and this PR would need a rebase**
- Resolve this PR first, and then, those PR would need a rebase

**Comment by saintplay** [comment]
**Finally we should fix the test suite**

First of all this syntax will pass everytime

```js
await wrapper.vm.$nextTick(() => {
  expect(true).toEqual(false)
})
// Passed
```

It should be

```js
await wrapper.vm.$nextTick()
return expect(true).toEqual(false)
// Failed
```

Second, there are test failing/passing depending on the timezone like this

image

**Comment by charliekassel** [comment]
Pretty sure that test doesn't need async await now that I'm using `vue-test-utils` as a mounting option.
Not sure what the best way to test dates across all timezones though, any suggestions?

**Comment by saintplay** [comment]
I updated the test with timezones issues.

**Comment by saintplay** [comment]
**[QUESTION]** There is a specific reason to add the **hours** and **minutes** in the `this.days`, `this.months`, `this.years` computeds?

image
 
Comparing dates will ignore **hours** and **minutes** anyway

**Comment by charliekassel** [comment]
It is an attempt to preserve the timezone of the users date

**Comment by saintplay** [comment]
I'm working on rebasing to 1.0.

**Comment by saintplay** [comment]
I just finished the rebase @charliekassel

File: src/utils/DateUtils.js [hunk 1]
```diff
@@ -2,6 +2,108 @@ import en from '../locale/translations/en'
...
+   * @param {Boolean} useUtc
+   */
+  setMonth (date, value, useUtc) {
+    return useUtc ? date.setUTCMonth(value) : date.setMonth(value)
+  },
+
+  /**
+   * Sets the date, using UTC or not
+   * @param {Date} date
+   * @param {Boolean} useUtc
```
Comment (charliekassel) [inline src/utils/DateUtils.js]: missing `@param {Number} value`

**Comment by charliekassel** [comment]
My general concern here is that the DateUtils functions are getting overly complex, at least in the calling of them by always passing in the `useUtc` flag as the last param to most of the functions.
I wonder whether if we made `DateUtils` a class with a `useUtc` property, rather than a collection of functions, we could keep it's api cleaner.
What do you think @saintplay ?
I generally feel that methods with boolean parameters are not too easy to read/understand and feel kind of hacky.

**Comment by saintplay** [comment]
Yes I understand, passing `this.useUTC` is super repetitive.

That's why I added `primiteGetFullYear`(should be called `internalGetFullYear`), because I didn't want to repeat the `this.useUTC` param. Since the rewrite to multiple components, this internal methods can't be applied.

I will suggest some kind of constructed wrapper for Date Utils too (like the class aproach).

---

```js
...
data () {
const constructedDateUtils = DateUtils.makeDateUtils(useUtc)
return {
...
utils: constructedDateUtils,
...
}
```

*and then to call the utc functions:*

```diff
- return DateUtils.getDay(dObj, this.useUtc)
+ return this.utils.getDay(dObj)
```

**Comment by saintplay** [comment]
Another approach, mentioned [here](https://www.monterail.com/blog/2016/simple-guide-to-authoring-open-source-vue-js-components) too is to create a **mixin** with these utilities date functions, and use it in `DateInput.vue`, `Datepicker.vue`, `PickerDay.vue`, etc.

And then to call the utc functions:

```diff
// Before

- DateUtils.getDay(dObj, this.useUtc)
```
---

```diff
// After

+ internalGetDay(dObj)
```
I would go with this approach, seems more clean and extensible.

@charliekassel

**Comment by charliekassel** [comment]
A mixin would have access to the instance properties and the one we are interested in `useUtc`, but it's methods would be added to each component which seems a bit repetitive and would bloat the compiled code a bit more than a utility library.
I personally prefer the factory function to return the utils class.

**Comment by saintplay** [comment]
@charliekassel done!

Hopefully this is the last update before merging it

**Comment by charliekassel** [comment]
Thanks @saintplay - I don't like the fact that you are appending the useUtc param to every method, that seem quite brittle to me.
I can't fork your fork so not sure how to add my work to your PR but this is more how I see it:
tldr: `useUtc` as a property of the utils object

```
import en from '../locale/translations/en'

const utils = {
  /**
   * @type {Boolean}
   */
  useUtc: false,
  /**
   * Returns the full year, using UTC or not
   * @param {Date} date
   */
  getFullYear (date) {
    return this.useUtc ? date.getUTCFullYear() : date.getFullYear()
  },

  /**
   * Returns the month, using UTC or not
   * @param {Date} date
   */
  getMonth (date) {
    return this.useUtc ? date.getUTCMonth() : date.getMonth()
  },

  /**
   * Returns the date, using UTC or not
   * @param {Date} date
   */
  getDate (date) {
    return this.useUtc ? date.getUTCDate() : date.getDate()
  },

  /**
   * Returns the day, using UTC or not
   * @param {Date} date
   */
  getDay (date) {
    return this.useUtc ? date.getUTCDay() : date.getDay()
  },

  /**
   * Returns the hours, using UTC or not
   * @param {Date} date
   */
  getHours (date) {
    return this.useUtc ? date.getUTCHours() : date.getHours()
  },

  /**
   * Returns the minutes, using UTC or not
   * @param {Date} date
   */
  getMinutes (date) {
    return this.useUtc ? date.getUTCMinutes() : date.getMinutes()
  },

  /**
   * Sets the full year, using UTC or not
   * @param {Date} date
   */
  setFullYear (date, value, useUtc) {
    return this.useUtc ? date.setUTCFullYear(value) : date.setFullYear(value)
  },

  /**
   * Sets the month, using UTC or not
   * @param {Date} date
   */
  setMonth (date, value, useUtc) {
    return this.useUtc ? date.setUTCMonth(value) : date.setMonth(value)
  },

  /**
   * Sets the date, using UTC or not
   * @param {Date} date
   * @param {Number} value
   */
  setDate (date, value, useUtc) {
    return this.useUtc ? date.setUTCDate(value) : date.setDate(value)
  },

  /**
   * Check if date1 is equivalent to date2, without comparing the time
   * @see https://stackoverflow.com/a/6202196/4455925
   * @param {Date} date1
   * @param {Date} date2
   */
  compareDates (date1, date2) {
    const d1 = new Date(date1.getTime())
    const d2 = new Date(date2.getTime())

    if (this.useUtc) {
      d1.setUTCHours(0, 0, 0, 0)
      d2.setUTCHours(0, 0, 0, 0)
    } else {
      d1.setHours(0, 0, 0, 0)
      d2.setHours(0, 0, 0, 0)
    }
    return d1.getTime() === d2.getTime()
  },

  /**
   * Validates a date object
   * @param {Date} date - an object instantiated with the new Date constructor
   * @return {Boolean}
   */
  isValidDate (date) {
    if (Object.prototype.toString.call(date) !== '[object Date]') {
      return false
    }
    return !isNaN(date.getTime())
  },

  /**
   * Return abbreviated week day name
   * @param {Date}
   * @param {Array}
   * @return {String}
   */
  getDayNameAbbr (date, days) {
    if (typeof date !== 'object') {
      throw TypeError('Invalid Type')
    }
    return days[this.getDay(date)]
  },

  /**
   * Return name of the month
   * @param {Number|Date}
   * @param {Array}
   * @return {String}
   */
  getMonthName (month, months) {
    if (!months) {
      throw Error('missing 2nd parameter Months array')
    }
    if (typeof month === 'object') {
      return months[this.getMonth(month)]
    }
    if (typeof month === 'number') {
      return months[month]
    }
    throw TypeError('Invalid type')
  },

  /**
   * Return an abbreviated version of the month
   * @param {Number|Date}
   * @return {String}
   */
  getMonthNameAbbr (month, monthsAbbr) {
    if (!monthsAbbr) {
      throw Error('missing 2nd paramter Months array')
    }
    if (typeof month === 'object') {
      return monthsAbbr[this.getMonth(month)]
    }
    if (typeof month === 'number') {
      return monthsAbbr[month]
    }
    throw TypeError('Invalid type')
  },

  /**
   * Alternative get total number of days in month
   * @param {Number} year
   * @param {Number} m
   * @return {Number}
   */
  daysInMonth (year, month) {
    return /8|3|5|10/.test(month) ? 30 : month === 1 ? (!(year % 4) && year % 100) || !(year % 400) ? 29 : 28 : 31
  },

  /**
   * Get nth suffix for date
   * @param {Number} day
   * @return {String}
   */
  getNthSuffix (day) {
    switch (day) {
      case 1:
      case 21:
      case 31:
        return 'st'
      case 2:
      case 22:
        return 'nd'
      case 3:
      case 23:
        return 'rd'
      default:
        return 'th'
    }
  },

  /**
   * Formats date object
   * @param {Date}
   * @param {String}
   * @param {Object}
   * @return {String}
   */
  formatDate (date, format, translation) {
    translation = (!translation) ? en : translation
    let year = this.getFullYear(date)
    let month = this.getMonth(date) + 1
    let day = this.getDate(date)
    let str = format
      .replace(/dd/, ('0' + day).slice(-2))
      .replace(/d/, day)
      .replace(/yyyy/, year)
      .replace(/yy/, String(year).slice(2))
      .replace(/MMMM/, this.getMonthName(this.getMonth(date), translation.months))
      .replace(/MMM/, this.getMonthNameAbbr(this.getMonth(date), translation.monthsAbbr))
      .replace(/MM/, ('0' + month).slice(-2))
      .replace(/M(?!a|ä|e)/, month)
      .replace(/su/, this.getNthSuffix(this.getDate(date)))
      .replace(/D(?!e|é|i)/, this.getDayNameAbbr(date, translation.days))
    return str
  },

  /**
   * Creates an array of dates for each day in between two dates.
   * @param {Date} start
   * @param {Date} end
   * @return {Array}
   */
  createDateArray (start, end) {
    let dates = []
    while (start <= end) {
      dates.push(new Date(start))
      start = this.setDate(new Date(start), this.getDate(new Date(start)) + 1)
    }
    return dates
  }
}

export const makeDateUtils = useUtc => Object.assign(utils, {useUtc})

export default {
  ...utils,
  makeDateUtils
}
// eslint-disable-next-line
;
```

**Comment by saintplay** [comment]
I agree! done!

**Comment by charliekassel** [comment]
🎉 💯 👍

**Comment by charliekassel** [comment]
Available in v1.4.0 - thanks @saintplay!
